### PR TITLE
Fix our tests on windows

### PR DESCRIPTION
**Pull request by iarna**
Making a PR for this to track our work. 

After cloning the repo make sure you do an `npm rebuild` in order to get cmdshims for your binaries.

Also be aware that `tap`'s glob support doesn't seem to support windows-type-slashes, so `npm run tap test/tap/l*.js` works but `npm run tap test\tap\l*.js` does not.

I've not tested test/tap/registry.js as I don't have couchdb in windows as yet.

Currently failing tests are: (as run from the "Github Shell", which is a powershell with git stuff in the path)
#### subshell issues
- [x] test/tap/builtin-config.js _fails when run from a path with a space in its name_
- [x] test/tap/verify-no-lifecycle-on-repo.js _(opened up a new cmd window‼)_
#### symlinks
- [x] test/tap/do-not-remove-other-bins.js
- [x] test/tap/gently-rm-linked-module.js _fails when run with `npm run tap`
- [x] test/tap/gently-rm-symlinked-global-dir.js
- [x] test/tap/lifecycle-path.js (yes, this one is in $PATH too)
- [x] test/tap/publish-access-unscoped-restricted-fails.js
- [x] test/tap/symlink-cycle.js
- [x] test/tap/unit-gentlyrm.js
#### unicode not defaulting to true
- [x] test/tap/extraneous-dep-cycle-ls-ok.js _fails only because we're not defaulting unicode to true_
- [x] test/tap/ls-l-depth-0.js _fails only because we're not defaulting unicode to true_
#### $PATH
- [x] test/tap/lifecycle-path.js _pmfi?_
#### $ var substitutions
- [x] test/tap/umask-lifecycle.js
#### config/env parsing issues

Usually just stuff with strings ending up as numbers, I think
- [x] test/tap/config-basic.js
- [x] test/tap/config-builtin.js
#### git-related
- [x] test/tap/add-remote-git-file.js
- [x] test/tap/github-shortcut.js
- [x] test/tap/version-update-shrinkwrap.js
- [x] test/tap/version-lifecycle.js
#### simple pathname matches
- [x] test/tap/dedupe-scoped.js
- [x] test/tap/outdated-depth-integer.js
- [x] test/tap/unit-child-path.js
- [x] test/tap/unit-module-name.js
- [x] test/tap/unit-package-id.js
#### maximum callstack exceeded :scream:
- [x] test/tap/legacy-\* _(fail for many reasons, not the least that mkdirp.sync is blowing the stack recursing: there may also be a Tacks bug)_
#### `\r\n` issues
- [x] test/tap/config-certfile.js (there looks like some `\r` chars are missing? in the cert as read?)
- [x] test/tap/graceful-restart.js (this one is a simple issue with `\r\n` in the test checks)
- [x] test/tap/outdated.js
#### Misc
- [x] test/tap/check-permissions.js
- [x] test/tap/config-edit.js _fails if there's a space in your path_
- [x] test/tap/cruft-test.js
- [x] test/tap/install-scoped-link.js _fails if there's a space in your path_
- [x] test/tap/git-cache-no-hooks.js _only fails with `npm t`_
- [x] test/tap/logout.js
- [x] test/tap/outdated-depth-deep.js _fails when run from a path with a space in its name_
#### ENOSYS kill
- [x] test/tap/publish-config.js
- [x] test/tap/unpublish-config.js
#### misc

These are either unclear problems, or ungrouped ones that don't seem to have other related ones.
- [x] test/tap/config-project.js
- [x] test/tap/git-cache-locking.js
- [x] test/tap/git-npmignore.js
- [x] test/tap/git-races.js _fails if there's a space in your path_
- [x] test/tap/install-bad-man.js
- [x] test/tap/install-link-scripts.js
- [x] test/tap/install-man.js
- [x] test/tap/install-save-local.js
- [x] test/tap/install-scoped-already-installed.js
- [x] test/tap/logout-scoped.js
- [x] test/tap/shrinkwrap-save-with-existing-dev-deps.js
- [x] test/tap/sorted-package-json.js
- [x] test/tap/spawn-enoent-help.js
- [x] test/tap/cache-shasum.js _fails if there's a space in your path_

### CODE REVIEW

**Comment by jdalton** [comment]
The WebApps and Frameworks team (my team) on Edge is looking into the symlinks issue this month.
Related to [win-symlink-test](https://github.com/iarna/win-symlink-test).

**Comment by jdalton** [comment]
We're investigating if [bash on Windows](https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/) will solve this issue.

**Comment by jdalton** [comment]
If not we'll contact the approp folks to get this on the right feature team's radar.

**Comment by jdalton** [comment]
\cc @AmazingJaze

**Comment by jdalton** [comment]
@iarna To help get the right folks at MS looking into this can you explain the impact that non-admin symlink support would have on npm. Is it perf, is it edge case support, etc?

**Comment by iarna** [comment]
@jdalton I'll write up something and CC you on Twitter with it.

**Comment by jdalton** [comment]
:sunglasses:

**Comment by iarna** [comment]
```
> tap test/tap/git-dependency-install-link.js
test/tap/git-dependency-install-link.js ............. 11/13 8s
  install from git repo [with --link]
  not ok error reading child package.json
    found:
      name: Error
      stack: "Error: ENOENT: no such file or directory, open 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json'\n    at Error (native)"
      message: "ENOENT: no such file or directory, open 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json'"
      errno: -4058
      code: ENOENT
      syscall: open
      path: "C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json"
    at:
      line: 92
      column: 9
      file: test/tap/git-dependency-install-link.js
    stack: >
      test/tap/git-dependency-install-link.js:92:9

      ReadFileContext.<anonymous>
      (node_modules/read-package-json/read-json.js:90:21)

      ReadFileContext.callback (node_modules/graceful-fs/graceful-fs.js:78:16)

      FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:324:13)
```

**Comment by iarna** [comment]
```
> tap test/tap/cache-shasum.js
test/tap/cache-shasum.js .............................. 4/5 2s
  cleanup
  not ok Error: ENOTEMPTY: directory not empty, rmdir 'C:\Users\Rebecca\Documents\GitHub\npm\test\tap\cache-shasum\request\2.27.0'
    at:
      native: true
      function: Error
    stack: |
      Object.fs.rmdirSync (fs.js:806:18)
      rmkidsSync (node_modules/rimraf/rimraf.js:334:11)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      rimrafSync (node_modules/rimraf/rimraf.js:295:9)
      node_modules/rimraf/rimraf.js:332:5
      rmkidsSync (node_modules/rimraf/rimraf.js:331:26)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      rimrafSync (node_modules/rimraf/rimraf.js:295:9)
      node_modules/rimraf/rimraf.js:332:5
      rmkidsSync (node_modules/rimraf/rimraf.js:331:26)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      Function.rimrafSync [as sync] (node_modules/rimraf/rimraf.js:295:9)
      test/tap/cache-shasum.js:51:10
    errno: -4051
    code: ENOTEMPTY
    syscall: rmdir
    path: "C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\cache-shasum\\request\\2.27.0"
    test: cleanup
    message: "Error: ENOTEMPTY: directory not empty, rmdir 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\cache-shasum\\request\\2.27.0'"
```

File: test/tap/git-npmignore.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-var cat = require('graceful-fs').writeFileSync
-var exec = require('child_process').exec
+var writeFileSync = require('graceful-fs').writeFileSync
```
Comment (xialeizhou) [inline test/tap/git-npmignore.js]: cool~

**Comment by zkat** [comment]
@othiym23 can I get a due diligence sheep?

**Comment by othiym23** [comment]
To record our in-air resolution: AppVeyor commits squished down to one, bug fixes left independent, try to make the test changes not too chunky but squash them down as well. :sheep::rocket::fire::tada:

**Comment by iarna** [comment]
Landed in 3.9.0!!!
